src/hugr.rs

```diff
@@ -38,15 +40,67 @@ pub struct Hugr {
     root: portgraph::NodeIndex,
 
     /// Operation types for each node.
-    op_types: UnmanagedDenseMap<portgraph::NodeIndex, OpType>,
+    op_types: UnmanagedDenseMap<portgraph::NodeIndex, NodeType>,
 
     /// Node metadata
     metadata: UnmanagedDenseMap<portgraph::NodeIndex, NodeMetadata>,
 }
 
+#[derive(Clone, Debug, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+/// The type of a node on a graph
+pub struct NodeType {
+    /// The underlying OpType
+    pub op: OpType,
+    /// The resources that the signature has been specialised to
+    pub input_resources: Option<ResourceSet>,
+}
+
+impl NodeType {
+    /// Create a new optype with some ResourceSet
+    pub fn new(op: impl Into<OpType>, input_resources: ResourceSet) -> Self {
+        NodeType {
+            op: op.into(),
+            input_resources: Some(input_resources),
+        }
+    }
+
+    /// Instantiate an OpType with no input resources
+    pub fn pure(op: impl Into<OpType>) -> Self {
+        NodeType {
+            op: op.into(),
+            input_resources: Some(ResourceSet::new()),
+        }
+    }
+
+    /// Instantiate an OpType with an unknown set of input resources
+    pub fn open_resources(op: impl Into<OpType>) -> Self {
+        NodeType {
+            op: op.into(),
+            input_resources: None,
+        }
+    }
+
+    /// Use the input resources to calculate the concrete signature of the node
+    pub fn signature(&self) -> Option<Signature> {
+        self.input_resources
+            .clone()
+            .map(|rs| self.op.signature().with_input_resources(rs))
+    }
+}
+
```

Comment:
I agree that it would reduce the size of this PR, but I didn't do it because it didn't seem right. Going this way requires forcing a semantics onto the `OpType` that seemed out of `into()`'s purview

I think the more neutral `Into` implementation might even be `NodeType::open_resources`?